rollup/sync_service.go

```diff
@@ -1059,8 +1082,11 @@ func (s *SyncService) ProcessSequencerBatchAppendedLog(ctx context.Context, ethl
 				gasPrice := new(big.Int).SetUint64(uint64(ethsign.gasPrice))
 				data := ethsign.data
 				l1BlockNumber := element.BlockNumber
-				// Set the L1TxOrigin to `nil`
-				tx = types.NewTransaction(nonce, to, big.NewInt(0), gasLimit, gasPrice, data, nil, l1BlockNumber, types.QueueOriginSequencer, types.SighashEthSign)
+				if to == (common.Address{}) {
+					tx = types.NewContractCreation(nonce, big.NewInt(0), gasLimit, gasPrice, data, nil, l1BlockNumber, types.QueueOriginSequencer)
+				} else {
+					tx = types.NewTransaction(nonce, to, big.NewInt(0), gasLimit, gasPrice, data, nil, l1BlockNumber, types.QueueOriginSequencer, types.SighashEthSign)
+				}
```

Comment:
Maybe this if/else should be refactored inside `NewTransaction` since you do it twice here and in L1052? Seems like a footgun to need to always check that when creating a new transaction and setting the sighash mode.

Reply:
I dislike the way that this codepath came out and would like to deduplicate it since its nearly the same for `CTCTransactionTypeEIP155` and `CTCTransactionTypeEthSign`. If we had logic inside of  `types.NewTransaction` that called `types.NewContractCreation` if the address was the zero address, we would just have to be careful about the queue origin sequencer transactions that come in via the RPC. We currently reject transactions sent to the zero address at the RPC layer, the implementation of `types.NewTransaction` would then depend on that